File: 1erdepartamental/python/15.recursividad_fractales.py
# ============================================
# PROGRAMA: Fractales con Recursividad
# DESCRIPCIÓN: Sierpinski y Cantor con recursión
# ============================================


class Sierpinski:
    """
    Triángulo de Sierpinski fractal.

    Cada nivel tiene 3^n triángulos
    """

    def __init__(self, nivel):
        self.__nivel = nivel

    @property
    def nivel(self):
        return self.__nivel

    def contar_triangulos(self, n=None):
        """
        T(n) = 3 × T(n-1)
        Base: T(0) = 1
        """
        if n is None:
            n = self.__nivel
        if n == 0:
            return 1
        return 3 * self.contar_triangulos(n - 1)

    def total_acumulado(self, n=None):
        """Suma de triángulos desde nivel 0 hasta n"""
        if n is None:
            n = self.__nivel
        if n == 0:
            return 1
        return self.contar_triangulos(n) + self.total_acumulado(n - 1)

    def dibujar(self, nivel=None):
        """Genera representación ASCII del fractal"""
        if nivel is None:
            nivel = self.__nivel
        if nivel == 0:
            return ["*"]

        t = self.dibujar(nivel - 1)
        h = len(t)

        # Superior: triángulo centrado
        resultado = [" " * h + linea for linea in t]
        # Inferior: dos copias
        for i in range(h):
            resultado.append(t[i] + " " + t[i])

        return resultado


class Cantor:
    """
    Conjunto de Cantor fractal.

    Cada nivel tiene 2^n segmentos
    """

    def __init__(self, nivel):
        self.__nivel = nivel

    @property
    def nivel(self):
        return self.__nivel

    def contar_segmentos(self, n=None):
        """
        S(n) = 2 × S(n-1)
        Base: S(0) = 1
        """
        if n is None:
            n = self.__nivel
        if n == 0:
            return 1
        return 2 * self.contar_segmentos(n - 1)

    def generar_segmentos(self, n, inicio=0, fin=81):
        """
        Genera posiciones de segmentos.
        Divide en 3, elimina tercio central
        """
        if n == 0:
            return [[inicio, fin]]

        tercio = (fin - inicio) / 3

        # Recursión en tercios izquierdo y derecho
        izq = self.generar_segmentos(n - 1, inicio, inicio + tercio)
        der = self.generar_segmentos(n - 1, inicio + 2 * tercio, fin)

        return izq + der


# ============================================
# EJECUCIÓN
# ============================================
if __name__ == "__main__":
    s = Sierpinski(4)
    print(f"Triángulos nivel 4: {s.contar_triangulos()}")
    print(f"Total acumulado: {s.total_acumulado()}")

    print("\nSierpinski nivel 3:")
    for linea in s.dibujar(3):
        print(linea)

    c = Cantor(4)
    print(f"\nSegmentos Cantor nivel 4: {c.contar_segmentos()}")

File: 1erdepartamental/c++/15.recursividad_fractales.cpp
// ============================================
// PROGRAMA: Fractales Recursivos
// DESCRIPCIÓN: Triángulo de Sierpinski y Conjunto de Cantor
// ============================================

#include <iostream>
#include <string>
using namespace std;


/**
 * Generador de fractales con recursión
 */
class FractalesRecursivos {
public:
    // Base: nivel == 0 -> imprime triángulo simple
    // Recursivo: dibuja 3 sub-triángulos más pequeños
    void sierpinski(int nivel, int tamano = 16) {
        if (nivel == 0) {
            dibujarTriangulo(tamano);
            return;
        }
        
        cout << "Sierpinski nivel " << nivel << ":" << endl;
        sierpinski(nivel - 1, tamano / 2);
        cout << endl;
    }
    
    // Base: nivel == 0 -> línea completa
    // Recursivo: divide en tercios, borra el del medio
    void cantor(int nivel, int longitud = 81) {
        if (nivel == 0) {
            dibujarLinea(longitud);
            cout << endl;
            return;
        }
        
        cantor(nivel - 1, longitud / 3);
        cout << string(longitud / 3, ' ');
        cantor(nivel - 1, longitud / 3);
    }
    
    void mostrarCantor(int niveles) {
        cout << "Conjunto de Cantor (nivel " << niveles << "):" << endl;
        cantorIterativo(0, niveles, 81);
    }

private:
    void dibujarTriangulo(int tamano) {
        for (int i = 0; i < tamano; i++) {
            cout << string(tamano - i - 1, ' ');
            cout << string(2 * i + 1, '*');
            cout << endl;
        }
    }
    
    void dibujarLinea(int longitud) {
        cout << string(longitud, '#');
    }
    
    // Versión iterativa para mejor visualización
    void cantorIterativo(int nivelActual, int nivelMax, int longitud) {
        if (nivelActual > nivelMax) return;
        
        if (nivelActual == 0) {
            dibujarLinea(longitud);
            cout << endl;
        } else {
            int seg = longitud / (1 << nivelActual);  // 2^nivelActual
            for (int i = 0; i < (1 << nivelActual); i++) {
                if (esSegmentoActivo(i, nivelActual)) {
                    cout << string(seg, '#');
                } else {
                    cout << string(seg, ' ');
                }
            }
            cout << endl;
        }
        
        cantorIterativo(nivelActual + 1, nivelMax, longitud);
    }
    
    bool esSegmentoActivo(int seg, int nivel) {
        // Un segmento está activo si no está en el tercio medio
        for (int i = 0; i < nivel; i++) {
            if ((seg % 3) == 1) return false;
            seg /= 3;
        }
        return true;
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    FractalesRecursivos fractales;
    
    fractales.sierpinski(1, 8);
    cout << endl;
    
    fractales.mostrarCantor(3);
    
    return 0;
}
